### ISSUE TITLE
Too many TCP connections established

### ISSUE THREAD

**Issue by yilezhu** [minio/minio-dotnet#411]
When I conduct the concurrent test of Minio, I find that if I upload files concurrently, a lot of TCP connections will be established and will not be released。
My test code is as follows:
`
static void Main(string[] args)
        {

            try
            {
                Run().Wait();
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
            }
            Console.ReadLine();
        }

        private static async Task Run()
        {
            // Make a new bucket called mymusic.
            var bucketName = "testsize"; //<==== change this
            var location = "us-east-1";
            var filePath = @"C:\Users\admin\Desktop\c143a9bc-be71-4475-9981-3c0d8f60d7f5";
            var contentType = "application/octet-stream";
            var endpoint = "192.168.1.14:9000";
            var accessKey = "minioadmin";
            var secretKey = "minioadmin";
            try
            {
                //FileStream fileStream = new FileStream(filePath, FileMode.Open, FileAccess.Read);
              
                var minio = new MinioClient(endpoint, accessKey, secretKey);

                bool found = await minio.BucketExistsAsync(bucketName);
                if (!found)
                {
                    await minio.MakeBucketAsync(bucketName, location);
                }

                Console.WriteLine("10万个文件上传开始~~~~" + DateTime.Now);
                Stopwatch stopwatch = Stopwatch.StartNew();
                await Task.WhenAll(Enumerable.Range(0, 100_000)
                .Select(_ => Task.Run(async () =>
                {
                    //fileStream.Position = 0;
                    await minio.PutObjectAsync(bucketName, Guid.NewGuid().ToString("N"), filePath, contentType);
                }))).ConfigureAwait(false);
                stopwatch.Stop();

                Console.WriteLine("10万个文件上传结束~~~~" + DateTime.Now);
                Console.WriteLine("总共用时" + stopwatch.ElapsedMilliseconds / 1000 / 60 + "分钟");
                //fileStream.Dispose();
            }
            catch (Exception e)
            {
                Console.WriteLine(e);
            }
        }
`
At this time, I run the test code and view the TCP link through the tcpview tool。
Found that the TCP connection has been established and has not been released。At this time, the CPU of my local computer has gone up to 100%。
![image](https://user-images.githubusercontent.com/18655233/81523981-3015fb80-9382-11ea-90f0-05d66298f78a.png)

By looking at the source code, it is found that when communicating, the application is made by csharprest, and a method that has been abandoned is used. I don't know if this is the reason。
the code like this
`
private async Task<IRestResponse> ExecuteTaskCoreAsync(IEnumerable<ApiResponseErrorHandlingDelegate> errorHandlers, IRestRequest request, CancellationToken cancellationToken = default(CancellationToken))
        {
           ...
//It's abandoned
            IRestResponse response = await this.restClient.ExecuteTaskAsync(request, cancellationToken).ConfigureAwait(false);
            ...
        }
`

**Comment by harshavardhana** [comment]
Are you sure you are making sure to close and drain the stream? 

I wonder because connection pooling is done externally and we have no control over it.

**Comment by yilezhu** [comment]
Yes, in fact, I have also tested the upload method through the file path, and the result is the same。
`
var minio = new MinioClient(endpoint, accessKey, secretKey);

                bool found = await minio.BucketExistsAsync(bucketName);
                if (!found)
                {
                    await minio.MakeBucketAsync(bucketName, location);
                }

                Console.WriteLine("10万个文件上传开始~~~~" + DateTime.Now);
                Stopwatch stopwatch = Stopwatch.StartNew();
                await Task.WhenAll(Enumerable.Range(0, 100_000)
                .Select(_ => Task.Run(async () =>
                {
                    await minio.PutObjectAsync(bucketName, Guid.NewGuid().ToString("N"), filePath, contentType);
                }))).ConfigureAwait(false);
`

**Comment by harshavardhana** [comment]
Interesting we shall investigate.. an open TCP Conn is not a problem it just shows connection pooling and keep-alive the question is does it keep growing and ultimately crash the system?

**Comment by yilezhu** [comment]
At present, I test that the TCP connection is growing all the time. When my local port is occupied, the test program will crash！

**Comment by yilezhu** [comment]
Just now, I did a single thread test and found that the same problem would occur. I would create a TCP link until the local port was occupied and then the program crashed！
the test code like this
```C#
   for (var a = 0; a < 100_000; a++)
                {
                    await minio.PutObjectAsync(bucketName, Guid.NewGuid().ToString("N"), filePath, contentType);
                }
```

**Comment by yilezhu** [comment]
When I use Parallel for testing, ithas also been building TCP links, and will report the following errors
![image](https://user-images.githubusercontent.com/18655233/81529742-17154680-9392-11ea-9990-36b44601adbb.png)
the test code like this
```
 Parallel.For(0, 100_000, new ParallelOptions { MaxDegreeOfParallelism = 4 }, async i =>
                {
                    await minio.PutObjectAsync(bucketName, Guid.NewGuid().ToString("N"), filePath, contentType);
                });
```

**Comment by yilezhu** [comment]
Hi ,I tested the `RestSharp` components separately, and found that the problem of TCP leakage was caused by the HTTP component used. I tried to submit an issue to the official, hoping to solve it!

**Comment by harshavardhana** [comment]
Fixed by #551
